Added linking between member and subscription created events

SimonBackx opened the pull request:
fixes https://github.com/TryGhost/Team/issues/2160

- Adds a `batch_id` to both events that contain the same ID if they were created at the same time.
- Removes duplicate signup/conversion events using the batch_id
- Requires an update in mongo-knex to work (refs https://ghost.slack.com/archives/C02G9E68C/p1666773313272409?thread_ts=1666767872.375009&cid=C02G9E68C)
- Some dependencies needed an update to load the latest mongo-knex
- Added tiers to membersUtils, loaded on startup (we can start to use this instead of fetching it every time)

Commit:
Added linking between member and subscription created events

refs https://github.com/TryGhost/Team/issues/2160

Adds a `request_id` to both events that contain the same ID if they were created at the same time.

Commit:
fixup! Added linking between member and subscription created events

allouis commented on ghost/core/core/server/data/migrations/versions/5.21/2022-10-26-04-50-member-subscription-created-request-id.js:
```diff
@@ -0,0 +1,73 @@
+
+            // Update the members_created_events table using INSERT ON DUPLICATE KEY UPDATE trick
+            const response1 = await knex('members_created_events').insert(batch.map((r) => {
+                return {
```
Do we need all of the columns here, or can we just have `id` and `request_id`?

allouis replied:
After testing I see we do need all the columns

SimonBackx replied:
Yup, MySQL complains that they don't have a default set if we don't send them

allouis commented on ghost/core/core/server/data/migrations/versions/5.21/2022-10-26-04-50-member-subscription-created-request-id.js:
```diff
@@ -0,0 +1,73 @@
+        const batchSize = 1000;
+    
+        for (let i = 0; i < rows.length; i += batchSize) { // eslint-disable-line no-restricted-syntax
+            const batch = rows.slice(i, i + batchSize);
```
We've previously used the `chunk` method from `lodash` instead of this - not a big deal tho

daniellockyer replied:
@SimonBackx Yeah I'd prefer to use `_.chunk` here if possible 😬

allouis reviewed (commented):
Looks good to me, I do wonder about the `request_id` naming though. Perhaps `transaction_id` or `group_id` or `batch_id` works better, as it is more specific to the concept of a batch of events, rather than a request 🤷

SimonBackx commented:
I like the `batch_id` naming! Wasn't a fan of `transaction_id` because that can be confused with SQL transactions

allouis commented:
Yeah agree RE transaction!

Commit:
Added signup/conversion grouping in activity feed

fixes https://github.com/TryGhost/Team/issues/2160

- Requires an update in mongo-knex to work (refs https://ghost.slack.com/archives/C02G9E68C/p1666773313272409?thread_ts=1666767872.375009&cid=C02G9E68C)
- Some dependencies needed an update to load the latest mongo-knex
- Added tiers to membersUtils, loaded on startup (we can start to use this instead of fetching it every time)

Commit:
Renamed request_id to batch_id

daniellockyer commented on ghost/core/core/server/data/migrations/versions/5.21/2022-10-26-04-50-member-subscription-created-request-id.js:
```diff
@@ -0,0 +1,73 @@
+
+module.exports = createTransactionalMigration(
+    async function up(knex) {
+        if (DatabaseInfo.isSQLite(knex)) {
```
Is there a specific reason we're skipping doing this on SQLite? Just curious

SimonBackx replied:
`TIMESTAMPDIFF` isn't supported on SQLite (we could probably search for a workaround), and I wasn't sure about 'conflict on insert' working properly over there. For non-production databases I couldn't really see a value in merging the events

daniellockyer commented on ghost/core/core/server/data/migrations/versions/5.21/2022-10-26-04-50-member-subscription-created-request-id.js:
```diff
@@ -0,0 +1,73 @@
+                return {
+                    id: r.m_id,
+                    batch_id: r.batch_id,
+                    member_id: r.member_id,
```
Small suggestion but can you comment the fields that you were forced to include because it's an INSERT vs the ones we actually care about (`id`, `batch_id`)?

Commit:
fixup! Added signup/conversion grouping in activity feed
